crates/holochain/src/core/workflow/app_validation_workflow/tests.rs

```diff
@@ -204,6 +205,309 @@ async fn main_loop_app_validation_workflow() {
+    let cell_id = app.cells()[0].cell_id().clone();
+
+    let app_validation_workspace = Arc::new(AppValidationWorkspace::new(
+        conductor
+            .get_or_create_authored_db(&dna_hash, cell_id.agent_pubkey().clone())
+            .unwrap()
+            .into(),
+        conductor.get_dht_db(&dna_hash).unwrap(),
+        conductor.get_dht_db_cache(&dna_hash).unwrap(),
+        conductor.get_cache_db(&cell_id).await.unwrap(),
+        conductor.keystore(),
+        Arc::new(dna_file.dna_def().clone()),
+    ));
+
+    // check there are no ops to app validate
+    // genesis entries have already been validated at this stage
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap()
+            .len();
+    assert_eq!(ops_to_validate, 0);
+
+    // insert create and delete op in dht db and mark ready for app validation
+    app_validation_workspace.dht_db.test_write(move |txn| {
+        insert_op(txn, &dht_delete_op_hashed).unwrap();
+        put_validation_limbo(txn, &dht_delete_op_hash, ValidationStage::SysValidated).unwrap();
+        insert_op(txn, &dht_create_op_hashed).unwrap();
+        put_validation_limbo(
+            txn,
+            &dht_create_op_hashed.hash,
+            ValidationStage::SysValidated,
+        )
+        .unwrap();
+    });
+
+    // check create and delete op are now counted as ops to validate
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap()
+            .len();
+    assert_eq!(ops_to_validate, 2);
+
+    let validation_dependencies = Arc::new(Mutex::new(ValidationDependencies::new()));
+
+    // run validation workflow
+    // outcome should be complete
+    let outcome_summary = app_validation_workflow_inner(
+        Arc::new(dna_hash.clone()),
+        app_validation_workspace.clone(),
+        conductor.raw_handle(),
+        &conductor.holochain_p2p().to_dna(dna_hash.clone(), None),
+        conductor
+            .get_or_create_space(&dna_hash)
+            .unwrap()
+            .dht_query_cache,
+        validation_dependencies.clone(),
+    )
+    .await
+    .unwrap();
+    assert_matches!(
+        outcome_summary,
+        OutcomeSummary {
+            ops_to_validate: 2,
+            validated: 2,
+            accepted: 2,
+            rejected: 0,
+            missing: 0,
+        }
+    );
+
+    // check ops to validate is also 0
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap()
+            .len();
+    assert_eq!(ops_to_validate, 0);
+}
+
+// test that app validation validates multiple ops in one workflow run where
+// one op depends on the other op
+// TODO this test only passes because actions are mistakenly written to the
```

Comment:
Did we decide that this is actually a mistake?

Reply:
No, it's not a mistake. It doesn't matter much, because the follow up PR deletes the TODO.

Reply:
I see what the difference is. Done https://github.com/holochain/holochain/pull/3732/files#diff-62b92dee463214ee3a6a703cb367099a8373bdd10265616c958020c04fca75acR248-R257